lib/pbio/drv/legodev/legodev_pup.c

```diff
@@ -314,62 +330,38 @@ static void debug_state_change(ext_dev_t *dev) {
 #define debug_state_change(...)
 #endif
 
-PROCESS_THREAD(pbio_legodev_pup_process, ev, data) {
-    static struct etimer timer;
+static PT_THREAD(pbdrv_legodev_pup_thread(ext_dev_t * dev)) {
 
-    static ext_dev_t *dev;
+    PT_BEGIN(&dev->pt);
 
-    PROCESS_BEGIN();
+    while (true) {
 
-    // Some hubs turn on power to the I/O ports in the bootloader. This causes
-    // UART sync delays after boot. It is faster to power cycle the I/O devices
-    // than it is to wait for long sync in most cases.
-    pbdrv_ioport_enable_vcc(false);
-    etimer_set(&timer, 500);
-    PROCESS_WAIT_EVENT_UNTIL(ev == PROCESS_EVENT_TIMER && etimer_expired(&timer));
-    pbdrv_ioport_enable_vcc(true);
-
-    etimer_set(&timer, 2);
-
-    while (!pbsys_status_test(PBIO_PYBRICKS_STATUS_SHUTDOWN)) {
-        PROCESS_WAIT_EVENT();
-        if (ev == PROCESS_EVENT_TIMER && etimer_expired(&timer)) {
-            etimer_reset(&timer);
-
-            for (int i = 0; i < PBDRV_CONFIG_LEGODEV_PUP_NUM_EXT_DEV; i++) {
-                dev = &ext_devs[i];
-
-                // If UART device was connected but not reporting data, restart detection.
-                if (dev->connected_type_id == PBDRV_LEGODEV_TYPE_ID_LPF2_UNKNOWN_UART &&
-                    dev->connected_type_id == dev->prev_type_id && timer_expired(&dev->dcm.watchdog)) {
-                    dev->connected_type_id = PBDRV_LEGODEV_TYPE_ID_NONE;
-                    dev->dcm.dev_id_match_count = 0;
-                }
-
-                // Keep running device detection unless UART sensor is attached.
-                if (dev->connected_type_id != PBDRV_LEGODEV_TYPE_ID_LPF2_UNKNOWN_UART) {
-                    poll_dcm(dev);
-                }
-
-                // Nothing more to do if it's the same device as before.
-                if (dev->connected_type_id == dev->prev_type_id) {
-                    continue;
-                }
+        // Initially assume nothing is connected.
+        dev->dcm.connected_type_id = PBDRV_LEGODEV_TYPE_ID_NONE;
+        dev->dcm.dev_id_match_count = 0;
 
+        // Run the device detection manager until a UART device is detected.
```

Comment:
It wouldn't hurt to add some more explanation here about the difference between UART devices and non-UART devices, in particular in relation to how disconnect is detected and why UART is a special case that stops the DCM but other devices don't.